core/embed/rust/src/ui/component/marquee.rs

```diff
@@ -0,0 +1,229 @@
+    pause_token: Option<TimerToken>,
+    min_offset: i16,
+    max_offset: i16,
+    state: State,
+    text: StrBuffer,
+    font: Font,
+    fg: Color,
+    bg: Color,
+    duration: Duration,
+    pause: Duration,
+}
+
+impl Marquee {
+    pub fn new(text: StrBuffer, font: Font, fg: Color, bg: Color) -> Self {
+        Self {
+            area: Rect::zero(),
+            pause_token: None,
+            min_offset: 0,
+            max_offset: 0,
+            state: State::Initial,
+            text,
+            font,
+            fg,
+            bg,
+            duration: Duration::from_millis(2000),
+            pause: Duration::from_millis(1000),
+        }
+    }
+
+    pub fn start(&mut self, ctx: &mut EventCtx, now: Instant) {
+        if let State::Initial = self.state {
+            let text_width = self.font.text_width(self.text.as_ref());
+            let max_offset = self.area.width() - text_width;
+
+            self.min_offset = 0;
+            self.max_offset = max_offset;
+
+            let anim = Animation::new(self.min_offset, max_offset, self.duration, now);
+
+            self.state = State::Left(anim);
+
+            // The animation is starting, request an animation frame event.
+            ctx.request_anim_frame();
+
+            // We don't have to wait for the animation frame event with the first paint,
+            // let's do that now.
+            ctx.request_paint();
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.state = State::Initial;
+    }
+
+    pub fn animation(&self) -> Option<&Animation<i16>> {
+        match &self.state {
+            State::Initial => None,
+            State::Left(a) => Some(a),
+            State::PauseLeft => None,
+            State::Right(a) => Some(a),
+            State::PauseRight => None,
+        }
+    }
+
+    pub fn is_at_right(&self, now: Instant) -> bool {
+        if let Some(p) = self.progress(now) {
+            return p == self.min_offset;
+        }
+        false
+    }
+
+    pub fn is_at_left(&self, now: Instant) -> bool {
+        if let Some(p) = self.progress(now) {
+            return p == self.max_offset;
+        }
+        false
+    }
+
+    pub fn progress(&self, now: Instant) -> Option<i16> {
+        self.animation().map(|a| a.value(now))
+    }
+
+    pub fn is_animating(&self) -> bool {
+        self.animation().is_some()
+    }
+
+    pub fn paint_anim(&mut self, offset: i16) {
+        display::marquee(
+            self.area,
+            self.text.as_ref(),
+            offset,
+            self.font,
+            self.fg,
+            self.bg,
+        );
+    }
+}
+
+impl Component for Marquee {
+    type Msg = Never;
+
+    fn place(&mut self, bounds: Rect) -> Rect {
+        let base_width = self.font.text_width("M");
+        let text_width = self.font.text_width(self.text.as_ref());
+        let area_width = bounds.width();
+
+        let shift_width = if area_width > text_width {
+            area_width - text_width
```

Comment:
When the text fits the area, is it going to just bounce from side to side? Maybe we should just not animate it in this case?

Reply:
i actually solved this outside of marquee like this (see `needs_marquee` flag): https://github.com/trezor/trezor-firmware/blob/grdddj/model_r_smaller/core/embed/rust/src/ui/model_tr/component/title.rs

Advantage is that if you actually want bouncing text you can use this, disadvantage is that you need to deal with that in every place where you don't want the bouncing text. 

Not sure what is preferable

Reply:
Can't imagine you'd ever want this, but maybe that's just problem with my imagination. I guess it really depends on the use cases ...